plugin/cmd/kube-scheduler/app/configurator.go

```diff
@@ -101,38 +102,91 @@ func createScheduler(
 	configurator = &schedulerConfigurator{
 		configurator,
 		s.PolicyConfigFile,
-		s.AlgorithmProvider}
+		s.AlgorithmProvider,
+		s.PolicyConfigMapName,
+		s.PolicyConfigMapNamespace,
+		s.UseLegacyPolicyConfig,
+	}
 
 	return scheduler.NewFromConfigurator(configurator, func(cfg *scheduler.Config) {
 		cfg.Recorder = recorder
 	})
 }
 
-// schedulerConfigurator is an interface wrapper that provides default Configuration creation based on user
-// provided config file.
+// schedulerConfigurator is an interface wrapper that provides a way to create
+// a scheduler from a user provided config file or ConfigMap object.
 type schedulerConfigurator struct {
 	scheduler.Configurator
-	policyFile        string
-	algorithmProvider string
+	policyFile               string
+	algorithmProvider        string
+	policyConfigMap          string
+	policyConfigMapNamespace string
+	useLegacyPolicyConfig    bool
 }
 
-// Create implements the interface for the Configurator, hence it is exported even through the struct is not.
-func (sc schedulerConfigurator) Create() (*scheduler.Config, error) {
-	if _, err := os.Stat(sc.policyFile); err != nil {
-		if sc.Configurator != nil {
-			return sc.Configurator.CreateFromProvider(sc.algorithmProvider)
+// getSchedulerPolicyConfig finds and decodes scheduler's policy config. If no
+// such policy is found, it returns nil, nil.
+func (sc schedulerConfigurator) getSchedulerPolicyConfig() (*schedulerapi.Policy, error) {
+	var configData []byte
+	var policyConfigMapFound bool
+	var policy schedulerapi.Policy
+
+	// If not in legacy mode, try to find policy ConfigMap.
+	if !sc.useLegacyPolicyConfig && len(sc.policyConfigMap) != 0 {
+		namespace := sc.policyConfigMapNamespace
+		policyConfigMap, err := sc.GetClient().CoreV1().ConfigMaps(namespace).Get(sc.policyConfigMap, metav1.GetOptions{})
+		if err != nil {
+			return nil, fmt.Errorf("Error getting scheduler policy ConfigMap: %v.", err)
+		}
+		if policyConfigMap != nil {
+			// We expect the first element in the Data member of the ConfigMap to
+			// contain the policy config.
+			if len(policyConfigMap.Data) != 1 {
```

Comment:
We are using an existing mechanism for deserializing both the contents of a ConfigMap entry and a legacy config file. Before calling deserialization we need to make this quick sanity check. I agree that the sanity check belongs to deserialization code, but since we are sharing the deserialization with config file, I thought it would be fine to perform the check here.